tutorial/tests/testsuite.py

```diff
@@ -49,9 +67,136 @@ def pytest_generate_tests(self, metafunc: pytest.Metafunc) -> None:
+    """Container class to store the test results when we collect them"""
+    stdout: str
+    stderr: str
+    test_name: str
+    success: bool
+
+
+@dataclasses.dataclass
+class OutputConfig:
+    """Container class to store the information to display in the test output"""
+    style: str
+    name: str
+    result: str
+
+
+def format_success_failure(syntax_error: bool, success: bool, name: str) -> OutputConfig:
+    """
+    Depending on the test results, returns a fragment that represents
+    either an error message, a success message, or a syntax error warning
+    """
+    if syntax_error:
+        return OutputConfig(
+            "alert-warning",
+            f"Tests <strong>CANNOT RUN</strong> for this cell.",
+            "&#129300 Careful, looks like you have a syntax error.",
+        )
+
+    if success:
+        return OutputConfig(
+            "alert-success",
+            f"Tests <strong>PASSED</strong> for the function <code>{name}</code>",
+            "&#x1F64C Congratulations, your solution was correct!",
+        )
+    else:
+        return OutputConfig(
+            "alert-danger",
+            f"Tests <strong>FAILED</strong> for the function <code>{name}</code>",
+            "&#x1F631 Your solution was not correct!",
+        )
+
+
+def format_long_stdout(text: str) -> str:
+    """
+    Format a long test stdout as a HTML by using the <details> element
+    """
+
+    def join_lines(text: str) -> str:
+        text = text.splitlines()
+        return "".join(
+            [f"<p>{line}</p>" for line in text[-6:-1]]
+        )
+
+    separator = 59 * ' _'
+    text = text.split(separator)[-1]
+    text_lines = "".join(join_lines(text))
+    test_runs = "".join(
+        [
+            f"""<details style="overflow-y: scroll; max-height: 200px;"><summary><u>Click here to expand</u></summary><div>{text_lines}</div></details></li>"""
+        ]
+    )
+    return test_runs
+
+
+class TestResultOutput(ipywidgets.VBox):
+    """Class to display the test results in a structured way"""
+
+    def __init__(self, name: str, syntax_error: bool, success: bool, test_outputs: List[TestResult]):
+        output_config = format_success_failure(syntax_error, success, name)
+        output_cell = ipywidgets.Output()
+
+        with output_cell:
+            custom_div = '<div style="border: 1px solid; border-color: lightgray; background-color: whitesmoke; margin: 5px; padding: 10px;">'
+            display(HTML("<h3>Test results</h3>"))
+            display(HTML(f"""<div class="alert alert-box {output_config.style}"><h4>{output_config.name}</h4>{output_config.result}</div>"""))
+
+            if not syntax_error:
+                if len(test_outputs) > 0 and test_outputs[0].stdout:
+                    display(HTML(f"<h4>Code output:</h4> {custom_div}{test_outputs[0].stdout}</div>"))
+
+                message = f"<h4>We tested your solution <code>solution_{name}</code> {len(test_outputs)} times with different inputs. "
```

Comment:
Could you generate this with string interpolation instead of concatenation? I find it easier to read